util/util.go

```diff
@@ -1,15 +1,82 @@
 package util
 
 import (
+	"context"
+	"errors"
 	"fmt"
 	"os"
 	"strings"
+	"time"
 	"unicode"
 	"unicode/utf8"
 
 	"github.com/algorand/go-codec/codec"
+
+	log "github.com/sirupsen/logrus"
 )
 
+// ErrTimeout is returned when CallWithTimeout has a normal timeout.
+var ErrTimeout = errors.New("timeout during call")
+
+// IsTimeoutError compares the given error against the timeout errors.
+func IsTimeoutError(err error) bool {
+	if err == nil {
+		return false
+	}
+	return errors.Is(err, ErrTimeout)
+}
+
+// ErrMisbehavingHandler is written to the log when a handler does not return.
+var ErrMisbehavingHandler = "Misbehaving handler did not exist after 1 second."
+
+// misbehavingHandlerDetector warn if ch does not exit after 1 second.
+func misbehavingHandlerDetector(log *log.Logger, ch chan struct{}) {
+	if log == nil {
+		return
+	}
+
+	select {
+	case <-ch:
+		// Good. This means the handler returns shortly after the context finished.
+		return
+	case <-time.After(1 * time.Second):
+		log.Warnf(ErrMisbehavingHandler)
+	}
+}
+
+// CallWithTimeout manages the channel / select loop required for timing
+// out a function using a WithTimeout context. No timeout if timeout = 0.
+// A new context is passed into handler, and cancelled at the end of this
+// call.
+func CallWithTimeout(ctx context.Context, log *log.Logger, timeout time.Duration, handler func(ctx context.Context) error) error {
+	if timeout == 0 {
+		return handler(ctx)
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// Call function in go routine
+	done := make(chan struct{})
+	var err error
+	go func(routineCtx context.Context) {
+		err = handler(routineCtx)
+		close(done)
+	}(timeoutCtx)
+
+	// wait for task to finish or context to timeout/cancel
+	select {
+	case <-done:
+		return err
```

Comment:
As discussed, before `return err` we should check if the context is expired and return 503 if so. Otherwise, on context expiration we can end up here and return a different error.